### ISSUE TITLE
Multiple and/or dynamic lists for comparison

### ISSUE THREAD

**Issue by tacotexmex** [dolanmiu/Minetest-Texture-Pack-Completion-Checker#1]
Using Summerfields file list as base for comparison has multiple problems: 
- All TPs miss/leave some textures to fallback. TPs compared to this will thus never be truly complete. Makes it harder to use Completion Checker as a contributor's tool.
- It supports several 3rd party mods. 100% is not achievable if TPs aren't supporting the exact same ones. Regardless, whatever support is present is not visible.

I propose we generate our own lists:

- Base list with all minetest PNGs (minus perhaps the android assets) that instead of "texture pack completion" says "Minetest Core"

- List with all minetest_game PNGs that instead of "texture pack completion" says "Minetest Game"

- Lists with PNGs (residing in textures folder only) for popular mods so that we can show percentage support for those as well. A little bit like [Summerfields website does](https://cloud.githubusercontent.com/assets/25328789/24863792/92d59a6c-1e02-11e7-89e4-9bc169c42c81.png) but automatic, good-looking and single download.

I don't know if you cache the Summerfields list in the service or if it's being fetched on request but it'd be very useful if both parts of comparison is switchable. 

For example: 
1. Compare list of PNGs inside `textures` folder in `minetest-mods/moreores` with the list of PNGs in `minetest-texturepacks/Good-Morning-Craft-Minetest`.
2. Put former repo name to use as badge label. (if possible strip all "minetest-" part if possible)

In the end we could have dynamic badges so that TP maintainers proudly could use them like so:
[Minetest Core | 100%]
[Minetest Game | 100%]
[3d_armor | 100%]
[more_ores | 82%]

**Comment by dolanmiu** [comment]
Could you make a list? You can post it here, or create a txt and upload it. I am not sure which files are technically needed!

I will convert it to JSON and make this project consume that instead of checking Summerfields

Thanks

**Comment by tacotexmex** [comment]
Alright!
### Minetest Core

> air.png
blank.png
camera_btn.png
chat_btn.png
debug_btn.png
down.png
down_arrow.png
drop_btn.png
fast_btn.png
fly_btn.png
gear_icon.png
halo.png
ignore.png
inventory_btn.png
jump_btn.png
left_arrow.png
logo.png
menu_bg.png
menu_header.png
minimap_mask_round.png
minimap_mask_square.png
minimap_overlay_round.png
minimap_overlay_square.png
no_screenshot.png
noclip_btn.png
object_marker_red.png
player_marker.png
progress_bar.png
progress_bar_bg.png
rangeview_btn.png
rare_controls.png
right_arrow.png
server_flags_creative.png
server_flags_damage.png
server_flags_favorite.png
server_flags_pvp.png
server_ping_1.png
server_ping_2.png
server_ping_3.png
server_ping_4.png
smoke_puff.png
sunrisebg.png
unknown_item.png
unknown_node.png
unknown_object.png
up_arrow.png

**Comment by tacotexmex** [comment]
### Minetest Game
> header.png
icon.png
beds_bed.png
beds_bed_fancy.png
beds_bed_foot.png
beds_bed_head.png
beds_bed_side1.png
beds_bed_side2.png
beds_bed_side_bottom.png
beds_bed_side_bottom_r.png
beds_bed_side_top.png
beds_bed_side_top_r.png
beds_bed_top1.png
beds_bed_top2.png
beds_bed_top_bottom.png
beds_bed_top_top.png
beds_transparent.png
boats_inventory.png
boats_wield.png
bones_bottom.png
bones_front.png
bones_rear.png
bones_side.png
bones_top.png
bucket.png
bucket_lava.png
bucket_river_water.png
bucket_water.png
carts_cart.png
carts_cart_front.png
carts_cart_side.png
carts_cart_top.png
carts_rail_crossing.png
carts_rail_crossing_brk.png
carts_rail_crossing_pwr.png
carts_rail_curved.png
carts_rail_curved_brk.png
carts_rail_curved_pwr.png
carts_rail_straight.png
carts_rail_straight_brk.png
carts_rail_straight_pwr.png
carts_rail_t_junction.png
carts_rail_t_junction_brk.png
carts_rail_t_junction_pwr.png
creative_trash_icon.png
character.png
bubble.png
crack_anylength.png
default_acacia_bush_stem.png
default_acacia_leaves.png
default_acacia_leaves_simple.png
default_acacia_sapling.png
default_acacia_tree.png
default_acacia_tree_top.png
default_acacia_wood.png
default_apple.png
default_aspen_leaves.png
default_aspen_sapling.png
default_aspen_tree.png
default_aspen_tree_top.png
default_aspen_wood.png
default_book.png
default_book_written.png
default_bookshelf.png
default_bookshelf_slot.png
default_brick.png
default_bronze_block.png
default_bronze_ingot.png
default_bush_stem.png
default_cactus_side.png
default_cactus_top.png
default_chest_front.png
default_chest_lock.png
default_chest_side.png
default_chest_top.png
default_clay.png
default_clay_brick.png
default_clay_lump.png
default_cloud.png
default_coal_block.png
default_coal_lump.png
default_cobble.png
default_copper_block.png
default_copper_ingot.png
default_copper_lump.png
default_coral_brown.png
default_coral_orange.png
default_coral_skeleton.png
default_desert_cobble.png
default_desert_sand.png
default_desert_sandstone.png
default_desert_sandstone_block.png
default_desert_sandstone_brick.png
default_desert_stone.png
default_desert_stone_block.png
default_desert_stone_brick.png
default_diamond.png
default_diamond_block.png
default_dirt.png
default_dry_grass.png
default_dry_grass_1.png
default_dry_grass_2.png
default_dry_grass_3.png
default_dry_grass_4.png
default_dry_grass_5.png
default_dry_grass_side.png
default_dry_shrub.png
default_fence_acacia_wood.png
default_fence_aspen_wood.png
default_fence_junglewood.png
default_fence_overlay.png
default_fence_pine_wood.png
default_fence_wood.png
default_flint.png
default_footprint.png
default_furnace_bottom.png
default_furnace_fire_bg.png
default_furnace_fire_fg.png
default_furnace_front.png
default_furnace_front_active.png
default_furnace_side.png
default_furnace_top.png
default_glass.png
default_glass_detail.png
default_gold_block.png
default_gold_ingot.png
default_gold_lump.png
default_grass.png
default_grass_1.png
default_grass_2.png
default_grass_3.png
default_grass_4.png
default_grass_5.png
default_grass_side.png
default_gravel.png
default_ice.png
default_iron_lump.png
default_item_smoke.png
default_junglegrass.png
default_jungleleaves.png
default_jungleleaves_simple.png
default_junglesapling.png
default_jungletree.png
default_jungletree_top.png
default_junglewood.png
default_key.png
default_key_skeleton.png
default_ladder_steel.png
default_ladder_wood.png
default_lava.png
default_lava_flowing_animated.png
default_lava_source_animated.png
default_leaves.png
default_leaves_simple.png
default_mese_block.png
default_mese_crystal.png
default_mese_crystal_fragment.png
default_mese_post_light_side.png
default_mese_post_light_side_dark.png
default_mese_post_light_top.png
default_meselamp.png
default_mineral_coal.png
default_mineral_copper.png
default_mineral_diamond.png
default_mineral_gold.png
default_mineral_iron.png
default_mineral_mese.png
default_mossycobble.png
default_obsidian.png
default_obsidian_block.png
default_obsidian_brick.png
default_obsidian_glass.png
default_obsidian_glass_detail.png
default_obsidian_shard.png
default_paper.png
default_papyrus.png
default_pine_needles.png
default_pine_sapling.png
default_pine_tree.png
default_pine_tree_top.png
default_pine_wood.png
default_rainforest_litter.png
default_rainforest_litter_side.png
default_river_water.png
default_river_water_flowing_animated.png
default_river_water_source_animated.png
default_sand.png
default_sandstone.png
default_sandstone_block.png
default_sandstone_brick.png
default_sapling.png
default_sign_steel.png
default_sign_wall_steel.png
default_sign_wall_wood.png
default_sign_wood.png
default_silver_sand.png
default_silver_sandstone.png
default_silver_sandstone_block.png
default_silver_sandstone_brick.png
default_snow.png
default_snow_side.png
default_snowball.png
default_steel_block.png
default_steel_ingot.png
default_stick.png
default_stone.png
default_stone_block.png
default_stone_brick.png
default_tool_bronzeaxe.png
default_tool_bronzepick.png
default_tool_bronzeshovel.png
default_tool_bronzesword.png
default_tool_diamondaxe.png
default_tool_diamondpick.png
default_tool_diamondshovel.png
default_tool_diamondsword.png
default_tool_meseaxe.png
default_tool_mesepick.png
default_tool_meseshovel.png
default_tool_mesesword.png
default_tool_steelaxe.png
default_tool_steelpick.png
default_tool_steelshovel.png
default_tool_steelsword.png
default_tool_stoneaxe.png
default_tool_stonepick.png
default_tool_stoneshovel.png
default_tool_stonesword.png
default_tool_woodaxe.png
default_tool_woodpick.png
default_tool_woodshovel.png
default_tool_woodsword.png
default_torch_animated.png
default_torch_on_ceiling_animated.png
default_torch_on_floor.png
default_torch_on_floor_animated.png
default_tree.png
default_tree_top.png
default_water.png
default_water_flowing_animated.png
default_water_source_animated.png
default_wood.png
gui_formbg.png
gui_furnace_arrow_bg.png
gui_furnace_arrow_fg.png
gui_hb_bg.png
gui_hotbar.png
gui_hotbar_selected.png
heart.png
player.png
player_back.png
wieldhand.png
doors_blank.png
doors_door_glass.png
doors_door_obsidian_glass.png
doors_door_steel.png
doors_door_wood.png
doors_item_glass.png
doors_item_obsidian_glass.png
doors_item_steel.png
doors_item_wood.png
doors_trapdoor.png
doors_trapdoor_side.png
doors_trapdoor_steel.png
doors_trapdoor_steel_side.png
dye_black.png
dye_blue.png
dye_brown.png
dye_cyan.png
dye_dark_green.png
dye_dark_grey.png
dye_green.png
dye_grey.png
dye_magenta.png
dye_orange.png
dye_pink.png
dye_red.png
dye_violet.png
dye_white.png
dye_yellow.png
farming_bread.png
farming_cotton.png
farming_cotton_1.png
farming_cotton_2.png
farming_cotton_3.png
farming_cotton_4.png
farming_cotton_5.png
farming_cotton_6.png
farming_cotton_7.png
farming_cotton_8.png
farming_cotton_seed.png
farming_desert_sand_soil.png
farming_desert_sand_soil_wet.png
farming_desert_sand_soil_wet_side.png
farming_flour.png
farming_soil.png
farming_soil_wet.png
farming_soil_wet_side.png
farming_straw.png
farming_tool_bronzehoe.png
farming_tool_diamondhoe.png
farming_tool_mesehoe.png
farming_tool_steelhoe.png
farming_tool_stonehoe.png
farming_tool_woodhoe.png
farming_wheat.png
farming_wheat_1.png
farming_wheat_2.png
farming_wheat_3.png
farming_wheat_4.png
farming_wheat_5.png
farming_wheat_6.png
farming_wheat_7.png
farming_wheat_8.png
farming_wheat_seed.png
fire_basic_flame.png
fire_basic_flame_animated.png
fire_flint_steel.png
flowers_dandelion_white.png
flowers_dandelion_yellow.png
flowers_geranium.png
flowers_mushroom_brown.png
flowers_mushroom_red.png
flowers_rose.png
flowers_tulip.png
flowers_viola.png
flowers_waterlily.png
flowers_waterlily_bottom.png
pbj_pup_back.png
pbj_pup_candies.png
pbj_pup_candies_animated.png
pbj_pup_front.png
pbj_pup_jelly.png
pbj_pup_sides.png
screwdriver.png
tnt_blast.png
tnt_boom.png
tnt_bottom.png
tnt_gunpowder_burning_crossing_animated.png
tnt_gunpowder_burning_curved_animated.png
tnt_gunpowder_burning_straight_animated.png
tnt_gunpowder_burning_t_junction_animated.png
tnt_gunpowder_crossing.png
tnt_gunpowder_curved.png
tnt_gunpowder_inventory.png
tnt_gunpowder_straight.png
tnt_gunpowder_t_junction.png
tnt_side.png
tnt_smoke.png
tnt_top.png
tnt_top_burning.png
tnt_top_burning_animated.png
vessels_drinking_glass.png
vessels_drinking_glass_inv.png
vessels_glass_bottle.png
vessels_glass_fragments.png
vessels_shelf.png
vessels_shelf_slot.png
vessels_steel_bottle.png
wool_black.png
wool_blue.png
wool_brown.png
wool_cyan.png
wool_dark_green.png
wool_dark_grey.png
wool_green.png
wool_grey.png
wool_magenta.png
wool_orange.png
wool_pink.png
wool_red.png
wool_violet.png
wool_white.png
wool_yellow.png
xpanes_bar.png
xpanes_bar_top.png
xpanes_pane_half.png
xpanes_space.png
xpanes_white.png

**Comment by tacotexmex** [comment]
For the record I just searched through https://github.com/minetest/minetest_game and https://github.com/minetest/minetest/tree/master/textures respectively. My command to do so after cloning the repos was `find . -type f -name "*.png" -execdir echo '{}' ';'`.

These are static list however. It would be cool if the checker itself could look into these repos for PNGs, but maintaining a list like this is _not_ hard work regardless. 🙂

**Comment by dolanmiu** [comment]
yes, the checker can check through the repos for PNGs. infact, github has an API which does it all for you, so it is very trivial.

So, do you think checking against https://github.com/minetest/minetest_game and https://github.com/minetest/minetest/tree/master/textures is a better idea than to use a static list? A static list isn't hard to maintain, but maintenance is still maintenance :P

**Comment by tacotexmex** [comment]
Yes, dynamically is a better idea. However they probably won't change paths for a good while. I vote for static maintained lists if the time saved on that could be put to better use, dynamic mod compatible file list checking for instance. But I realize that may be the same thing, code-wise.

**Comment by dolanmiu** [comment]
@tacotexmex 

It doesn't take very long to do it dynamically (it is done dynamically now). All i need to do is make a few tweaks to point to the two repos mentioned.

I have a few ideas on how to make it work for different mod packs.

something like:

```
https://minetest-checker.herokuapp.com/<MOD_NAME>/<OWNER>/<REPO>
```

Omitting the `mod_name` will just do the compare with `minetest_game` and the minetest textures. 

Then you can have multiple badges, one for mod completion, and one for minetest completion or something.

How does that sound? What are your thoughts?

**Comment by tacotexmex** [comment]
Sounds good generally but I wonder: Is `<MOD_NAME>` for being able to drill down to individual mods in modpacks? If so I'd rather have it look for `modpack.txt` first and adapt the search for multiple `textures` one directory level further down if it does exist.

**Comment by dolanmiu** [comment]
now i am confused, so each modpack contain a number of mods?

are we checking the entire modpack, or individual mods within in the modpack?

**Comment by tacotexmex** [comment]
A modpack is a collection of mods, yes. For a modpack to work in Minetest, an empty `modpack.txt` file needs to sit in the root directory of the pack. Therefore we can look for that file to know we're dealing with a pack.

So, for modpacks: multiple directories in root possibly contain a `textures` directory. For regular mods: `textures` directory sits in root.

So before running the file list fetching routine you can check for `modpack.txt` first, then run one of two routines since the directory structure looks a little different. Or if scanning for PNGs in `textures` directories in any part of a repo is cheap code wise, just do that. It should end up in a flat file list anyway, even for modpacks.

**Comment by tacotexmex** [comment]
Are we closer to achieving new list comparison? @dolanmiu :slightly_smiling_face:

**Comment by dolanmiu** [comment]
Not yet! Been quite busy this Easter break, busy today too, will have some free time on Monday though

**Comment by dolanmiu** [comment]
Can you give me a few examples of mod packs and mods in the meantime, so I can look around

**Comment by tacotexmex** [comment]
Sure thing!

These are modpacks
[mesecons](https://github.com/minetest-mods/mesecons)
[technic](https://github.com/minetest-mods/technic)
[pipeworks](https://github.com/minetest-mods/pipeworks)
[animals_modpack](https://github.com/sapier/animals_modpack)

These are regular mods
[farming](https://github.com/tenplus1/farming)
[awards](https://github.com/minetest-mods/awards)
[nether](https://github.com/PilzAdam/nether)

Happy Easter! 🐥

**Comment by dolanmiu** [comment]
So when checking a **mod pack**, will it by default check all mods within the mod pack, or can people choose which mod within the mod pack they can check?

**Comment by dolanmiu** [comment]
I have done it for the `farming` mod pack, check the updated readme for more info :) 

More mods will come soon

**Comment by tacotexmex** [comment]
> will it by default check all mods within the mod pack

Yes.

**Comment by tacotexmex** [comment]
Okay, I was hoping that the mod file list also could be pulled dynamically. It will be a nearly impossible to keep up with all mods?

**Comment by dolanmiu** [comment]
Just added support for `nether` and `awards`, check the `README` for more info.

> Okay, I was hoping that the mod file list also could be pulled dynamically. It will be a nearly impossible to keep up with all mods?

you are right, it will be a huge pain to keep up with all of it. I'm still working on it :P

I am looking at Summerfields: https://github.com/SummerFields/Summerfields-Minetest

And I see no indicator that it is using any modpacks, I am trying to find `technic`, `mescons`, `pipeworks`, and it seems like its not there

**Comment by dolanmiu** [comment]
@tacotexmex What other **mods** are popular that would require to have a checker? So far, this checker has `farming`, `awards` and `nether`

**Comment by tacotexmex** [comment]
Great work on Dynamic Mod Checking! Now there's just the modpack support left I guess? 😃

**Comment by dolanmiu** [comment]
@tacotexmex Yes dynamic left :) but I'm unsure on how a texture pack implements a mod pack, does summerfields or good morning craft have modpack support? I did a search and cannot find anything inside summerfields which suggests that it uses mod packs, unless I am being super dumb lol

**Comment by tacotexmex** [comment]
In texturepacks, modpacks are supported in the same way as normal mods are, there's no difference. For better or worse, texturepacks are completely flat structured. So there's no problem on that end.

If you check for `modpack.txt` in a repo, **you'll know absolutely that it's a modpack**, since the MT engine needs it.

The file structure, in comparison:

### Normal mod
```
/textures
(init.lua + other files)
```
### Modpack
```
/mod1
  /textures
/mod2
  /textures
/mod3
  /textures
modpack.txt (+ other files)
```
So check for `modpack.txt` to decide if it's a modpack or not.
